zerver/tornado/event_queue.py

```diff
@@ -968,7 +968,24 @@ def is_sender_client(client: ClientDescriptor) -> bool:
 
     for user_data in users:
         user_profile_id: int = user_data["id"]
-        flags: Collection[str] = user_data.get("flags", [])
+        flags: List[str] = user_data.get("flags", [])
+
+        # TODO/compatibility: During server upgrade if the event queue has
+        # events with the deprecated `wildcard_mentioned` flag, we need to
+        # update the `flags` to avoid error while handling such events moving forward.
+        # Remove this when one can no longer directly upgrade from 7.x to main.
+        if "wildcard_mentioned" in flags:
+            updated_flags = [
+                "stream_wildcard_mentioned" if flag == "wildcard_mentioned" else flag
+                for flag in flags
+            ]
+            flags = updated_flags
+
+        # TODO/compatibility: This logic is intended to support returning
+        # the `wildcard_mentioned` flag to older clients. Remove this when
+        # one can no longer directly upgrade from 7.x to main.
+        if "stream_wildcard_mentioned" in flags:
+            flags.append("wildcard_mentioned")
```

Comment:
I didn't use `if "stream_wildcard_mentioned" in flags or "topic_wildcard_mentioned" in flags:` because:

If we plan to return `wildcard_mentioned = topic_wild... | stream_wild...` then, suppose during upgrade we have an event in a queue with `wildcard_mentioned` flag, now in the translation code (just above) we should replace `wildcard_mentioned` with `stream_wild..` or `topic_wild...`?

Technically it is possible to set `wildcard_mentioned` flag using @**topic**, but the feature is not yet used by web client or mobile; we also haven't mentioned it yet in our api docs. So I think it is safe to move forward with the logic that `wildcard_mentioned` should be returned only if `stream_wildcard_mentioned` is set?

For older clients  `wildcard_mentioned` is only set in the case of  stream wildcard mentions as they are unaware of "topic mentions" feature?

Reply:
I think we want `if "stream_wildcard_mentioned" in flags or "topic_wildcard_mentioned" in flags`. The upgrade code path concern that you mention is generally going to affect clients for at most ~30 minutes after an upgrade -- so as long as nothing crashing, the impact of the behavior being non-ideal is very bounded, because clients will eventually reload and get a new event queue (see `server_generation` for details). If we sometimes produce an incorrect value for the flags temporarily after someone used the still-undocumented `@topic` mention code, it's not a big problem.

Whereas anyone using an old copy of the mobile apps that predates `topic_wildcard_mentioned` support, might see newly sent `@topic` messages as not containing a mention, without this code, unless I'm missing something?

Reply:
I could be wrong, but I think with the present proposed API design, we may not need any compatibility code in Tornado at all -- we just need to make sure that any Django code that is generating events containing a `flags` list adds the `wildcard_mentioned` while it's adding the other flags.

Reply:
Yeah, passing the three fields via send_event, instead of including wildcard_mentioned in tornado code is a much cleaner approach.